### PR TITLE
#252: Add set charts branch refs command

### LINKED ISSUE #252: Create a way to automate being able to set the dev or production branch for all Chart references
Current Manual Process:

Create a PR that moves from development branches to production branches for charts and system-charts.

**Update variables in settings/settings.go** 

- [ChartDefaultBranch](https://github.com/rancher/rancher/blob/release/v2.8/pkg/settings/setting.go#L117) - Branch where charts are pointed to 

**Update variables in scripts/package-env** - Used to properly generate the rancher-images.txt and rancher-image-sources.txt that are release artifacts 
- [SYSTEM_CHART_DEFAULT_BRANCH](https://github.com/rancher/rancher/blob/release/v2.8/scripts/package-env#L5) - Branch where system-charts are pointed to 
- [CHART_DEFAULT_BRANCH](https://github.com/rancher/rancher/blob/release/v2.8/scripts/package-env#L7) - Branch where charts are pointed to

**Update arguments in package/Dockerfile** 
- [SYSTEM_CHART_DEFAULT_BRANCH](https://github.com/rancher/rancher/blob/release/v2.8/package/Dockerfile#L19) - Branch where system-charts are pointed to
- [CHART_DEFAULT_BRANCH](https://github.com/rancher/rancher/blob/release/v2.8/package/Dockerfile#L20) - Branch where charts are pointed to

### PR DESCRIPTION
**Pull request by Tashima42**
Adds a command to update charts branch references in rancher
Closes #252

## How to test
* ```
    git switch 252-set-chart-references
   ```
* ```
    make rancher_release
    ```
* Fork [github.com/rancher/rancher](github.com/rancher/rancher)
Without PR creation:
* ```
    cmd/rancher_release/bin/rancher_release set-charts-branch-refs --fork-path YOUR-RANCHER-FORK --base-branch release/v2.8 --current-charts-branch dev-v2.8 --new-charts-branch TEST
    ```
With PR creation, if you remove `dry-run` it will push changes to your fork and try to create a PR in rancher/rancher
* ```
    export GITHUB_TOKEN=your-token
    ```
* ```
    cmd/rancher_release/bin/rancher_release set-charts-branch-refs --fork-path YOUR-RANCHER-FORK --base-branch release/v2.8 --current-charts-branch dev-v2.8 --new-charts-branch TEST --create-pr --fork-owner YOUR-GH-USERNAME --dry-run
    ```

### CODE REVIEW

File: release/rancher/rancher.go [hunk 1]
```diff
@@ -27,50 +27,69 @@ const (
...
-	RancherRepoDir    string
-	CurrentKDMBranch  string
-	NewKDMBranch      string
+type SetBranchReferencesArgs struct {
+	RancherRepoPath   string
+	CurrentBranch     string
+	NewBranch         string
 	RancherBaseBranch string
 	DryRun            bool
+	BranchName        string
```
Comment (briandowns) [inline release/rancher/rancher.go:92]: Can you move this up so the bool is the last field?

File: release/rancher/rancher.go [hunk 2]
```diff
@@ -183,60 +202,97 @@ func rancherHelmChartVersions(repoURL string) ([]string, error) {
...
 
 func SetKDMBranchReferences(ctx context.Context, forkPath, rancherBaseBranch, currentKDMBranch, newKDMBranch, forkOwner, githubToken string, createPR, dryRun bool) error {
-	if _, err := os.Stat(forkPath); err != nil {
-		return err
-	}
-	scriptPath := filepath.Join(forkPath, setKDMBranchReferencesScriptFile)
-	f, err := os.Create(scriptPath)
-	if err != nil {
-		return err
+	branchName := "kdm-set-" + newKDMBranch
```
Comment (briandowns) [inline release/rancher/rancher.go:205]: Can this be a const?

File: release/rancher/rancher.go [hunk 3]
```diff
@@ -27,50 +27,69 @@ const (
 	rancherImagesFileName    = "/rancher-images.txt"
 	rancherHelmRepositoryURL = "https://releases.rancher.com/server-charts/latest/index.yaml"
 
-	setKDMBranchReferencesScriptFile = "set_kdm_branch_references.sh"
-	setKDMBranchReferencesScript     = `#!/bin/bash
+	setKDMBranchReferencesScriptFileName = "set_kdm_branch_references.sh"
+	setChartReferencesScriptFileName     = `set_chart_references_script.sh`
+	cloneCheckoutRancherScript           = `#!/bin/bash
```
Comment (briandowns) [inline release/rancher/rancher.go:32]: Can you run an example of this script through shellcheck and remediate anything it finds?

File: release/rancher/rancher.go (hunk 2 above)
Comment (Tashima42) [reply release/rancher/rancher.go:205]: No, because `newKDMBranch` is not a const

File: release/rancher/rancher.go [hunk 4]
```diff
@@ -27,49 +24,80 @@ const (
 	rancherImagesFileName    = "/rancher-images.txt"
 	rancherHelmRepositoryURL = "https://releases.rancher.com/server-charts/latest/index.yaml"
 
-	setKDMBranchReferencesScriptFile = "set_kdm_branch_references.sh"
-	setKDMBranchReferencesScript     = `#!/bin/bash
-set -x
+	setKDMBranchReferencesScriptFileName = "set_kdm_branch_references.sh"
+	setChartReferencesScriptFileName     = `set_chart_references.sh`
+	cloneCheckoutRancherScript           = `#!/bin/bash
```
Comment (briandowns) [inline release/rancher/rancher.go:29]: Are we using bash specific functionality? If not, can you change this to `#!/bin/sh`? Shellcheck should confirm POSIX compliance.

**Review by briandowns — APPROVED** [review]

**Review by luisrojas2 — APPROVED** [review]
LGTM

**Pull request merged**